File: samples/app-complete-auth/csharp/AppCompleteAuth/Dialogs/BotSsoAuthDialog.cs
using AdaptiveCards;
using AppCompleteAuth.Models;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Graph;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace AppCompleteAuth.Dialogs
{
    /// <summary>
    /// BotSsoAuthDialog class to handle Single Sign-On (SSO) authentication.
    /// </summary>
    public class BotSsoAuthDialog : LogoutDialog
    {
        private readonly ConcurrentDictionary<string, Token> _token;

        public BotSsoAuthDialog(string configuration, ConcurrentDictionary<string, Token> token)
            : base(nameof(BotSsoAuthDialog), configuration)
        {
            _token = token;
            AddDialog(new OAuthPrompt(
                nameof(OAuthPrompt),
                new OAuthPromptSettings
                {
                    ConnectionName = ConnectionName,
                    Text = "Please login",
                    Title = "Sign In",
                    Timeout = 1000 * 60 * 1, // User has 1 minute to login
                }));

            AddDialog(new TextPrompt(nameof(TextPrompt)));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
            {
                    PromptStepAsync,
                    LoginStepAsync,
                    UserInfoCardAsync
            }));

            InitialDialogId = nameof(WaterfallDialog);
        }

        /// <summary>
        /// Prompts the user to login.
        /// </summary>
        private async Task<DialogTurnResult> PromptStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            return await stepContext.BeginDialogAsync(nameof(OAuthPrompt), null, cancellationToken);
        }

        /// <summary>
        /// Handles the login step.
        /// </summary>
        private async Task<DialogTurnResult> LoginStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var tokenResponse = (TokenResponse)stepContext.Result;

            if (tokenResponse != null)
            {
                await stepContext.Context.SendActivityAsync(MessageFactory.Text("Login successful"), cancellationToken);
                var token = new Token
                {
                    AccessToken = tokenResponse.Token
                };

                _token.AddOrUpdate("token", token, (key, newValue) => token);
                return await stepContext.PromptAsync(
                    nameof(TextPrompt),
                    new PromptOptions
                    {
                        Prompt = MessageFactory.Text("What is your preferred user name?"),
                    },
                    cancellationToken);
            }

            return await stepContext.EndDialogAsync();
        }

        /// <summary>
        /// Displays the user information card.
        /// </summary>
        private async Task<DialogTurnResult> UserInfoCardAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var userName = stepContext.Result as string;
            _token.TryGetValue("token", out var token);
            var client = new SimpleGraphClient(token.AccessToken);
            var me = await client.GetMeAsync();
            var title = !string.IsNullOrEmpty(me.JobTitle) ? me.JobTitle : "Unknown";
            var photo = await client.GetPhotoAsync();
            var reply = MessageFactory.Attachment(GetProfileCard(me, photo, title, userName));
            await stepContext.Context.SendActivityAsync(reply, cancellationToken);
            return await stepContext.EndDialogAsync();
        }

        /// <summary>
        /// Creates a user profile card.
        /// </summary>
        private Microsoft.Bot.Schema.Attachment GetProfileCard(User profile, string photo, string title, string userName)
        {
            var card = new AdaptiveCard(new AdaptiveSchemaVersion(1, 0))
            {
                Body = new List<AdaptiveElement>
                    {
                        new AdaptiveTextBlock
                        {
                            Text = "User Information",
                            Size = AdaptiveTextSize.Default
                        },
                        new AdaptiveColumnSet
                        {
                            Columns = new List<AdaptiveColumn>
                            {
                                new AdaptiveColumn
                                {
                                    Items = new List<AdaptiveElement>
                                    {
                                        new AdaptiveImage
                                        {
                                            Url = new Uri(photo),
                                            Size = AdaptiveImageSize.Medium,
                                            Style = AdaptiveImageStyle.Person
                                        }
                                    },
                                    Width = "auto"
                                },
                                new AdaptiveColumn
                                {
                                    Items = new List<AdaptiveElement>
                                    {
                                        new AdaptiveTextBlock
                                        {
                                            Text = $"Hello! {profile.DisplayName}",
                                            Weight = AdaptiveTextWeight.Bolder,
                                            IsSubtle = true
                                        }
                                    },
                                    Width = "stretch"
                                }
                            }
                        },
                        new AdaptiveFactSet
                        {
                            Separator = true,
                            Facts = new List<AdaptiveFact>
                            {
                                new AdaptiveFact { Title = "Job title:", Value = title },
                                new AdaptiveFact { Title = "Email:", Value = profile.UserPrincipalName },
                                new AdaptiveFact { Title = "User name:", Value = userName }
                            }
                        }
                    }
            };

            return new Microsoft.Bot.Schema.Attachment
            {
                ContentType = AdaptiveCard.ContentType,
                Content = card
            };
        }
    }
}

File: samples/app-complete-auth/nodejs/dialogs/botSSOAuthDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { WaterfallDialog, TextPrompt } = require('botbuilder-dialogs');
const { CardFactory, MessageFactory, InputHints } = require('botbuilder');
const { LogoutDialog } = require('./logoutDialog');
const { SsoOAuthPrompt } = require('./ssoOAuthPrompt');
const { SimpleGraphClient } = require('../simpleGraphClient');

const SSOAUTH = 'SsoAuth';
const OAUTH_PROMPT = 'OAuthPrompt';
const TEXT_PROMPT = 'textPrompt';
let token;

/**
 * BotSSOAuthDialog class handles the SSO authentication process.
 */
class BotSSOAuthDialog extends LogoutDialog {
    constructor(id) {
        super(id, process.env.connectionName);

        this.addDialog(new SsoOAuthPrompt(OAUTH_PROMPT, {
            connectionName: process.env.connectionName,
            text: 'Please Sign In',
            title: 'Sign In',
            timeout: 300000
        }));
        this.addDialog(new TextPrompt(TEXT_PROMPT));
        this.addDialog(new WaterfallDialog(SSOAUTH, [
            this.promptStep.bind(this),
            this.loginStep.bind(this),
            this.userInfoStep.bind(this)
        ]));
        this.initialDialogId = SSOAUTH;
    }

    /**
     * Initiates the OAuth prompt.
     * @param {Object} stepContext - The context of the current step.
     */
    async promptStep(stepContext) {
        return await stepContext.beginDialog(OAUTH_PROMPT);
    }

    /**
     * Handles the login step after the OAuth prompt.
     * @param {Object} stepContext - The context of the current step.
     */
    async loginStep(stepContext) {
        if (stepContext.context._activity.text === "sso") {
            const tokenResponse = stepContext.result;

            if (!tokenResponse || !tokenResponse.token) {
                await stepContext.context.sendActivity('Login was not successful, please try again.');
            } else {
                token = tokenResponse.token;
                const messageText = 'What is your user name?';
                const msg = MessageFactory.text(messageText, messageText, InputHints.ExpectingInput);
                await stepContext.context.sendActivity('Login successful.');
                return await stepContext.prompt(TEXT_PROMPT, { prompt: msg });
            }
            return await stepContext.endDialog();
        }
    }

    /**
     * Retrieves and displays user information after login.
     * @param {Object} stepContext - The context of the current step.
     */
    async userInfoStep(stepContext) {
        const userName = stepContext.result;
        const client = new SimpleGraphClient(token);
        let imageString = "";
        let img2 = "";

        try {
            const myDetails = await client.getMeAsync();
            if (myDetails) {
                const userImage = await client.getUserPhoto();
                const result = await userImage.arrayBuffer();
                console.log(userImage.type);
                imageString = Buffer.from(result).toString('base64');
                img2 = `data:image/png;base64,${imageString}`;

                const userCard = CardFactory.adaptiveCard(this.getAdaptiveCardUserDetails(myDetails, img2, userName));
                await stepContext.context.sendActivity({ attachments: [userCard] });
            }
        } catch (error) {
            console.log(error);
        }

        return await stepContext.endDialog();
    }

    /**
     * Generates an adaptive card with user details.
     * @param {Object} myDetails - The user's details.
     * @param {string} userImage - The user's image in base64 format.
     * @param {string} userName - The user's name.
     * @returns {Object} - The adaptive card JSON.
     */
    getAdaptiveCardUserDetails(myDetails, userImage, userName) {
        return {
            $schema: 'http://adaptivecards.io/schemas/adaptive-card.json',
            body: [
                {
                    type: "TextBlock",
                    size: "Medium",
                    weight: "Bolder",
                    text: "User profile details are"
                },
                {
                    type: "Image",
                    size: "Medium",
                    url: userImage
                },
                {
                    type: "TextBlock",
                    size: "Medium",
                    weight: "Bolder",
                    wrap: true,
                    text: `Hello! ${myDetails.displayName}`
                },
                {
                    type: "TextBlock",
                    size: "Medium",
                    weight: "Bolder",
                    text: `Job title: ${myDetails.jobDetails ? myDetails.jobDetails : "Unknown"}`
                },
                {
                    type: "TextBlock",
                    size: "Medium",
                    weight: "Bolder",
                    text: `Email: ${myDetails.userPrincipalName}`
                },
                {
                    type: "TextBlock",
                    size: "Medium",
                    weight: "Bolder",
                    text: `User name: ${userName}`
                }
            ],
            type: 'AdaptiveCard',
            version: '1.4'
        };
    }
}

exports.BotSSOAuthDialog = BotSSOAuthDialog;